### DISCUSSION TITLE
Nuxt 3 SSR does execute XHR on client side and not server side

### DISCUSSION THREAD

**Discussion by FrazeColder** [nuxt/nuxt#20575] [Questions] [answered]
I have started to build a Nuxt 3 project. One of the big advantages of Nuxt is that it supports SSR out of the box, which is a big plus for SEO. This is because the whole page gets rendered on the server side and the user, as well as the Googlebot, gets served a fully rendered page with all the data. 

However, I am unable to set up such an infrastructure. I have coded my API backend with PHP and the Laravel framework. This is running via valet on the domain `backend.test`. Now I want my Nuxt 3 app to make a request to the following API endpoint `https://backend.test/api/v1/products` and how a list of all products being retrieved by the API. 

This also works, but the point is that the user gets served an empty HTML file instead of a server side rendered HTML file which already contains the products. So, the client is actually executing the XHR request to the API endpoint. But that's not what I want to archive. I want the page to be rendered by the server. 

This is my `List.vue` file, which should return an HTML list of all products:
```
<script setup>
const { data: products } = useFetch(
  "https://backend.test/api/v1/products?order_by=newest&expired=1&type[0]=DEAL",
  {
    key: Date.now().toString(),
    server: true,
  }
);
</script>

<template>
  <div>
    <p>Product List</p>
    {{ products }}
    <!--
    <div v-if="pending" class="grid grid-cols-4 gap-5">
      <div v-for="product in products">
        <div>{{ product }}</div>
      </div>
    </div>
    <div v-else>Loading!</div>
    -->
  </div>
</template>
```

As you can see, I also have added the parameter `server: true` which is true by default to tell Nuxt explicit that he should render it on the server, but he still doesn't. 

My `index.vue` file
```
<template>
  <div>
    <h1>Index</h1>
    <p>Index page</p>
    <List />
  </div>
</template>
```

My `app.vue` file:
```
<template>
  <div>
    <NuxtLayout>
      <NuxtPage />
    </NuxtLayout>
  </div>
</template>
```

My `nuxt.config.ts` file:
```
export default defineNuxtConfig({
  modules: [
    "@nuxtjs/eslint-module",
    "@nuxtjs/tailwindcss",
    "@pinia/nuxt",
    "@nuxt/devtools",
    "@nuxtjs/html-validator",
  ],
  css: ["~/assets/css/main.scss"],
  postcss: {
    plugins: {
      tailwindcss: {},
      autoprefixer: {},
    },
  },
});
```

I also added `ssr: true` to my `nuxt.config.ts` file, which is already true by default, but I still see the client making the XHR request.

What am I doing wrong here?

And I also have a second question:
When I remove `key: Date.now().toString(),` from my `useFetch` my data from the API is not being loaded at every page refresh but rather after like 5–10 seconds waiting. Why is that?

Kind regards

**Comment by warflash** [comment]
You are not awaiting the useFetch hence the server response does not contain the data. Awaiting should fix that.
 Apart from that in the unused code you're v-if checking if the response is pending and if it is pending rendering a list which is the wrong way round. You'd want to do `!pending`. Using the date as a key might not do you a favor either. Both of those things shouldn't matter for the issue but figured I would mention it

**Reply by FrazeColder** [reply]
I have changed my code to the following:

```
<script setup>
const { data: products, pending } = await useFetch(
  "https://backend.test/api/v1/products?order_by=newest&expired=1&type[0]=DEAL",
  {
    key: Date.now().toString(),
    server: true,
  }
);
</script>

<template>
  <div>
    <p>Product List</p>
    <div v-if="!pending" class="grid grid-cols-4 gap-5">
      <div v-for="product in products">
        <div>{{ product }}</div>
      </div>
    </div>
    <div v-else>Loading!</div>
  </div>
</template>
```

But sadly, it is still rendering on the client site and my HTML file which is being sent by the server is still not containing the product data.

**Reply by warflash** [reply]
That seems odd, could you try to provide a reproduction here? https://stackblitz.com/github/nuxt/starter/tree/v3/?file=README.md

**Reply by FrazeColder** [reply]
I just tried and unfortunately, with the following code it is working, but only on an empty repo and on stackblitz.com

(Just copy and paste the following code into app.vue):
```
<script setup>
const { data: products, pending } = await useFetch(
  "https://dummyjson.com/products",
  {
    key: Date.now().toString(),
    transform: (products) => products.products,
  }
);
</script>

<template>
  <div>
    <p>Product List</p>
    <div v-if="!pending" class="grid grid-cols-4 gap-5">
      <div v-for="product in products">
        <div>{{ product }}</div>
      </div>
    </div>
    <div v-else>Loading!</div>
  </div>
</template>
```

So, I will continue debugging and check what's the reason for. Maybe a third-party package?

**Reply by FrazeColder** [reply]
One more question: When I now execute the script above, the data is present in the HTML content.

However, when loading the page, I shortly see the data, then shortly the "Loading!" text and then the data again. 
Is there any way to tell `useFetch` to not load the data again on the client side but only when I press a refresh button, for example?

**Reply by warflash** [reply]
> Maybe a third-party package?

Seems very unlikely to me, I'd recommend trying to revert all the config changes etc. you tried and if that doesnt work then slowly start stripping unrelated stuff until you found whats causing it.

> However, when loading the page, I shortly see the data, then shortly the "Loading!" text and then the data again.

Remove the key property or set it to something other than the current date and it'll be fixed. Manual refreshing can be done by calling the refresh function provided in the return object from useFetch as well but not needed in this example

**Comment by maou-shonen** [comment]
It sounds like you're misunderstanding how nuxt3 renders pages?
[document rendering](https://nuxt.com/docs/guide/concepts/rendering)

**Comment by FrazeColder** [comment, marked as answer]
Finally, I have found what was causing my error. As I wrote, I have coded my backend in Laravel. My API is running with valet on a `.test` TLD with a self-signed HTTPS certificate provided by valet. 

Because the certificate verification by Node is unsuccessful, the server does not render the data and send an HTML document with the populated data. It simply refuses it. 

I was able to fix this problem by adding the following line to my `.env` file in my Nuxt 3 project:
```
NODE_TLS_REJECT_UNAUTHORIZED = 0
```
